firedrake/adjoint_utils/function.py

```diff
@@ -225,23 +225,32 @@ def _ad_convert_riesz(self, value, options=None):
         from firedrake import Function, Cofunction
 
         options = {} if options is None else options
-        riesz_representation = options.get("riesz_representation", "l2")
+        riesz_representation = options.get("riesz_representation", "L2")
         solver_options = options.get("solver_options", {})
         V = options.get("function_space", self.function_space())
 
-        if riesz_representation != "l2" and not isinstance(value, Cofunction):
-            raise TypeError("Expected a Cofunction")
-        elif not isinstance(value, (Number, Cofunction, Function)):
+        if not isinstance(value, (Number, Cofunction, Function, Number)):
             raise TypeError("Expected a Cofunction, Function or a float")
 
-        if riesz_representation == "l2":
-            if isinstance(value, (Cofunction, Function)):
-                return Function(V, val=value.dat)
-            else:
+        if isinstance(value, Number):
+            if value == 0.:
+                # l2 Riesz map is directly applied when the value is a real number 0..
+                # This is seen in adjoint-based derivative when the functional
+                # is independent of the control variable.
```

Comment:
I suggest the following comment
# In adjoint-based differentiation, the case isinstance(value, Number) and value == 0. arises
# only when the functional is independent on the control variable. In this case, we apply the
# l2 Riesz map